Start with `cfg_version` parsing

mo8it opened the pull request:
### What does this PR try to resolve?

Fixes #15531

mo8it commented:
@epage I will of course take care of proper error handling, but I just want to get feedback on the initial direction.

`cfg(nightly)` can be already parsed as a `Name`.

weihanglo reviewed (commented):
I just tweaked the PR description a bit. We usually keep a tracking issue open until stabilization.

mo8it commented on src/cargo/core/compiler/custom_build.rs:
```diff
@@ -416,6 +416,7 @@ fn build_work(build_runner: &mut BuildRunner<'_, '_>, unit: &Unit) -> CargoResul
                 let values = cfg_map.entry(k.as_str()).or_default();
                 values.push(v.as_str());
             }
+            Cfg::Version(..) => todo!(),
```
No idea what to do here

epage replied:
I think the following code will be relevant:
https://github.com/rust-lang/cargo/blob/47c911e9e6f6461f90ce19142031fe16876a3b95/src/cargo/ops/cargo_compile/mod.rs#L495-L511

mo8it replied:
The code you point to wasn't helpful. I still don't know in what format I should write the version cfg as an env var.

ehuss replied:
Just to be clear here, we do not expect rustc to emit this in `--print`, correct? That means this code path should never be reached, right?

And, similar to the point made at https://github.com/rust-lang/rust/pull/141137#discussion_r2096067981, it doesn't really make sense to include this here since it would need to emit *every* version to be correct. My inclination is that the env var shouldn't be set for build scripts.

epage replied:
Ah, yes, I missed what this code was doing. I was giving code to help when evaluating dependencies.  I agree with Eric that we shouldn't do anything here.  Maybe worth a `trace!`

epage commented on crates/cargo-platform/src/cfg.rs:
```diff
@@ -235,6 +250,25 @@ impl<'a> Parser<'a> {
                 Ok(CfgExpr::Not(Box::new(e)))
             }
+            Some(Ok(Token::Ident(false, "version"))) => {
+                self.t.next();
+                self.eat(&Token::LeftParen)?;
```
Can you have `--cfg version=hello` and so need to support `#[cfg(version)1.30.0)]` and `#[cfg(version = "hello")]`?

mo8it replied:
Looking at the updated RFC, I assume the answer is no?

epage replied:
I thought the stabilization report said that `--cfg version=hello` would still work

> #[cfg(version = "1.20.0")] is not supported, and there is a warning of the unexpected_cfgs lint (but no compilation error)

(not supported within the context of this feature, `unexpected_cfgs` firing because `--check-cfg` wasn't specified for `version`)

> #[cfg(version = "1.20.0")] acts as if the cfg was false due to the wrong syntax, even if the compiler version is above the specified version

(again, framed within this feature; it could be true if `--cfg version="1.20.0"` was specified)

mo8it replied:
I pushed a change so that `version = …` is ignored by the resolver now. I also raise a warning on it.

In what context could `--cfg version="1.23.4"` be used? I never specified a `cfg` flag. Where is that parsed? How should I handle it?

epage replied:
> In what context could --cfg version="1.23.4" be used? I never specified a cfg flag. Where is that parsed? How should I handle it?

More so the point is that this change should be agnostic to `--cfg version="1.23.4"` and should not change the behavior of `cfg(version = "1.23.4")`.

epage commented on crates/cargo-platform/src/cfg.rs:
```diff
@@ -398,6 +432,43 @@ impl<'a> Iterator for Tokenizer<'a> {
                 }
+                // Reject a major version starting with 0
+                Some((start, '1'..='9')) => {
+                    while let Some(&(end, ch)) = self.s.peek() {
+                        if ch == ')' {
+                            let mut iter = self.orig[start..end].split('.');
+                            let major = iter.next().unwrap().parse().unwrap();
+                            if let Some(minor) = iter.next().map(|s| s.parse().unwrap()) {
+                                if let Some(patch) = iter.next().map(|s| s.parse().unwrap()) {
+                                    if iter.next().is_some() {
```
This is some fairly deep nesting.  Is there an alternative?

epage commented on crates/cargo-platform/src/cfg.rs:
As a stylistic thing, as you add tests, we prefer tests to be added in a commit before the change, with them passing.  Your change then includes updating the tests to make sure they pass.  How to apply this is dependent on the PR (and sometimes it just doesn't work).  In this case, I'd write the tests without `#[cfg(version)]` and then add it in the follow ups.

epage commented on src/cargo/core/compiler/custom_build.rs:
```diff
@@ -416,6 +416,7 @@ fn build_work(build_runner: &mut BuildRunner<'_, '_>, unit: &Unit) -> CargoResul
                 let values = cfg_map.entry(k.as_str()).or_default();
                 values.push(v.as_str());
             }
+            Cfg::Version(..) => todo!(),
```
Should we have this behind an unstable gate or insta-stabilize once the stabilization report for the rustc side is done?

epage commented:
Be sure to check out rust-lang/rust#141137 for a summary of what has changed from the RFC

epage commented on crates/cargo-platform/src/cfg.rs:
```diff
@@ -124,30 +145,40 @@ impl fmt::Display for Cfg {
         match *self {
-            CfgExpr::Not(ref e) => !e.matches(cfg),
-            CfgExpr::All(ref e) => e.iter().all(|e| e.matches(cfg)),
-            CfgExpr::Any(ref e) => e.iter().any(|e| e.matches(cfg)),
+            CfgExpr::Not(ref e) => !e.matches(cfg, rustc_version),
+            CfgExpr::All(ref e) => e.iter().all(|e| e.matches(cfg, rustc_version)),
+            CfgExpr::Any(ref e) => e.iter().any(|e| e.matches(cfg, rustc_version)),
+            CfgExpr::Value(Cfg::Version(CfgRustVersion { minor, patch })) => {
+                match minor.cmp(&rustc_version.minor) {
+                    Ordering::Less => true,
+                    Ordering::Equal => patch <= rustc_version.patch,
+                    Ordering::Greater => false,
+                }
+            }
```
I believe the tracking issue mentions specific behavior for nightlies
- `1.80` includes 1.80 nightlies
- `1.80.0` does not include 1.80 nightlies

ehuss replied:
I'm not quite following this. Can you say more about where that is being discussed? That doesn't seem to be the current behavior.

epage replied:
Maybe I'm misinterpeting what was said in

> Ultimately, a compromise was agreed upon, in which nightly releases are treated as "complete" i.e. cfg(version(1.20)) evals to true on nightly builds with version 1.20.0, but there is a nightly flag -Z assume-incomplete-release to opt into the behaviour that doesn't do this assumption. This compromise was implemented in PR https://github.com/rust-lang/rust/pull/81468.

The actual rustc tests should be checked to confirm.  We probably should most or all of their tests.

mo8it replied:
Did I get the desired logic here?

https://github.com/rust-lang/cargo/blob/120bea75161530b7e4d3c26c1d5da32c95d740e3/crates/cargo-platform/tests/test_cfg.rs#L224-L234

https://github.com/rust-lang/cargo/blob/120bea75161530b7e4d3c26c1d5da32c95d740e3/crates/cargo-platform/src/cfg.rs#L36-L52

mo8it replied:
@epage Any update on that?

I will have time on Tuesday or Wednesday to finish the PR after clarifying the desired behavior for nightly.

epage commented on crates/cargo-platform/src/cfg.rs:
```diff
@@ -124,30 +145,40 @@ impl fmt::Display for Cfg {
         match *self {
             Cfg::Name(ref s) => s.fmt(f),
             Cfg::KeyPair(ref k, ref v) => write!(f, "{} = \"{}\"", k, v),
+            Cfg::Version(CfgRustVersion { minor, patch }) => {
+                write!(f, "version(\"1.{minor}.{patch}\")")
```
Can we delegate the formatting of `CfgRustVersion` to `CfgRustVersion`?

epage commented on crates/cargo-platform/src/cfg.rs:
```diff
@@ -124,30 +145,40 @@ impl fmt::Display for Cfg {
+            CfgExpr::Not(ref e) => !e.matches(cfg, rustc_version),
+            CfgExpr::All(ref e) => e.iter().all(|e| e.matches(cfg, rustc_version)),
+            CfgExpr::Any(ref e) => e.iter().any(|e| e.matches(cfg, rustc_version)),
+            CfgExpr::Value(Cfg::Version(CfgRustVersion { minor, patch })) => {
+                match minor.cmp(&rustc_version.minor) {
+                    Ordering::Less => true,
+                    Ordering::Equal => patch <= rustc_version.patch,
+                    Ordering::Greater => false,
+                }
+            }
```
Like we did with `package.rust-version`, could we delegate the matching logic to `CfgRustVersion`?

epage commented on crates/cargo-platform/src/lib.rs:
```diff
@@ -88,15 +88,26 @@ impl Platform {
+                                 https://doc.rust-lang.org/cargo/reference/features.html"
+                            ));
+                        },
+                        "version" => {
+                            warnings.push(String::from(
+                                "Found `version = ...` in `target.'cfg(...)'.dependencies`. \
+                                This format is ignored by Cargo. \
+                                Use the format `version(\"1.23.4\")` or `version(\"1.23\")` instead."
+                            ));
+                        },
```
Based on https://github.com/rust-lang/cargo/pull/15533#discussion_r2093685172 I don't think this should be present.

epage commented on crates/cargo-platform/tests/test_cfg.rs:
```diff
@@ -152,33 +172,34 @@ fn cfg_expr_bad() {
 
 #[test]
 fn cfg_matches() {
```
Can we add tests for `version`?

epage commented on tests/testsuite/cfg.rs:
```diff
@@ -817,3 +817,145 @@ fn cfg_booleans_rustflags_no_effect() {
         .run();
 }
+
+#[cargo_test]
+fn cfg_version() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "a"
+                edition = "2015"
+
+                [target.'cfg(version("1.87.0"))'.dependencies]
+                b = { path = 'b' }
+            "#,
+        )
+        .file("src/lib.rs", "extern crate b;")
+        .file("b/Cargo.toml", &basic_manifest("b", "0.0.1"))
+        .file("b/src/lib.rs", "")
+        .build();
+    p.cargo("check -v").run();
+}
```
I wish we had a way to override the rust version so we could test with `1.2345` (the version we always use for "far future" in tests)

epage replied:
Could we use `RUSTC_FORCE_RUSTC_VERSION`?

weihanglo commented on crates/cargo-platform/src/cfg.rs:
```diff
@@ -15,13 +16,33 @@ pub enum CfgExpr {
+
 /// A cfg value.
 #[derive(Eq, PartialEq, Hash, Ord, PartialOrd, Clone, Debug)]
 pub enum Cfg {
```
Should we make it non-exhaustive so we don't need to bump version in the future?

epage replied:
Keeping it this way makes it easier to find places to update, so unless we have a strong reason for avoiding major bumps, I'd lean towards this (similar with `cargo-util-schemas`)

mo8it commented on crates/cargo-platform/tests/test_cfg.rs:
```diff
@@ -152,33 +172,34 @@ fn cfg_expr_bad() {
-    assert!(e!(r # foo).matches(&[c!(foo)]));
-    assert!(e!(r # foo).matches(&[c!(r # foo)]));
+    let rustc_version = semver::Version::new(0, 0, 0);
+    assert!(e!(foo).matches(&[c!(bar), c!(foo), c!(baz)], &rustc_version));
```
Rewriting the Git history would be too tricky for me. It is my first real contribution to Cargo. I am learning to adapt the style and will apply it in future commits :)
Thanks for the feedback, much appreciated :D

epage replied:
I'm prepping a rewrite of the history for you which also includes switching `cargo-platform` over to snapshot testing

mo8it replied:
All edits are welcome <3

epage replied:
My commits are pushed.  Hope I didn't lose anything

mo8it commented:
@epage Thanks for rewriting the history. I found it a bit disappointing that I wasn't listed as an author anymore. Therefore, I updated the author information of the commits and hope that you are happy with the current state.

I also fixed the CI by removing `dbg!(input)`.

Finally, I resolved the merge conflicts.

Learning `jj` today helped a lot :D

If we agree on https://github.com/rust-lang/cargo/pull/15533#discussion_r2098745262, then the PR should be ready for merging.

epage commented:
> I found it a bit disappointing that I wasn't listed as an author anymore.

Sorry that I forgot to preserve that!

ehuss commented on tests/testsuite/cfg.rs:
```diff
@@ -817,3 +817,144 @@ fn cfg_booleans_rustflags_no_effect() {
+        .with_stderr_data(str![[r#"
+[LOCKING] 1 package to latest compatible version
+[CHECKING] foo v0.0.0 ([ROOT]/foo)
+error[E0463]: can't find crate for `b`
+ --> src/lib.rs:1:1
+  |
+1 | extern crate b;
+  | ^^^^^^^^^^^^^^^ can't find crate
+
+For more information about this error, try `rustc --explain E0463`.
+[ERROR] could not compile `foo` (lib) due to 1 previous error
```
We can't match on the exact output from rustc because if they change the output it would prevent this test from passing. I think something like the following should be close enough:

```suggestion
error[E0463]: can't find crate for `b`
...
[ERROR] could not compile `foo` (lib) due to 1 previous error
```

ehuss reviewed (changes requested):
Can you include some tests as mentioned from https://github.com/rust-lang/cargo/pull/15533#issuecomment-2895074734? Perhaps something like:

```rust
#[cargo_test]
fn cfg_version_from_index() {
    // Checks that cfg(version) works in the index.
    Package::new("only_newer", "1.0.0").publish();
    Package::new("bar", "1.0.0")
        .file("src/lib.rs", "extern crate only_newer;")
        .target_dep("only_newer", "1.0", "cfg(version(\"1.87\"))")
        .publish();

    let p = project()
        .file(
            "Cargo.toml",
            r#"
                [package]
                name = "foo"
                edition = "2024"

                [dependencies]
                bar = "1.0"
            "#,
        )
        .file("src/lib.rs", "")
        .build();

    p.cargo("check")
        .with_stderr_data(str![[r#"
[UPDATING] `dummy-registry` index
[LOCKING] 2 packages to latest Rust [..] compatible versions
[DOWNLOADING] crates ...
[DOWNLOADED] only_newer v1.0.0 (registry `dummy-registry`)
[DOWNLOADED] bar v1.0.0 (registry `dummy-registry`)
[CHECKING] only_newer v1.0.0
[CHECKING] bar v1.0.0
[CHECKING] foo v0.0.0 ([ROOT]/foo)
[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s

"#]])
        .run();

    p.cargo("tree")
        .with_stdout_data(str![[r#"
foo v0.0.0 ([ROOT]/foo)
└── bar v1.0.0
    └── only_newer v1.0.0

"#]])
        .run();
}

#[cargo_test]
fn cfg_version_from_index_in_future() {
    // Checks that cfg(version) for a version from the future is ignored.
    Package::new("only_newer", "1.0.0").publish();
    Package::new("bar", "1.0.0")
        .file("src/lib.rs", "")
        .target_dep("only_newer", "1.0", "cfg(version(\"1.2345\"))")
        .publish();

    let p = project()
        .file(
            "Cargo.toml",
            r#"
                [package]
                name = "foo"
                edition = "2024"

                [dependencies]
                bar = "1.0"
            "#,
        )
        .file("src/lib.rs", "")
        .build();

    p.cargo("tree")
        .with_stdout_data(str![[r#"
foo v0.0.0 ([ROOT]/foo)
└── bar v1.0.0

"#]])
        .run();
}
```

epage commented:
FYI a concern was raised on the new stabilization issue (https://github.com/rust-lang/rust/pull/141766#issuecomment-2946372974) about what syntax is used for this.

mo8it commented:
@epage @ehuss Unfortunately, I was not able to find time to finish this PR. I will also be away for about 3 weeks. So feel free to add your changes and then merge.

epage commented:
@mo8it don't worry, there is nothing to do atm while the design is being re-evaluated

ranger-ross commented:
Given that this has sat stale for over a year and the original issue is [`S-blocked-external`](https://github.com/rust-lang/cargo/issues?q=sort%3Aupdated-desc+state%3Aopen+label%3A%22S-blocked-external%22+) I am gonna go ahead and close this. 

Would be great to see this feature land in some form one day.
Thanks for your effort on this
